scripts/dts/flatten.py

```diff
@@ -0,0 +1,203 @@
+            root['props'] = devicetree.update_node(root['props'],
+                                                   value['props'])
+            return
+
+        root['props']['phandle'] = state['phandle']
+        state['phandle'] += 1
+
+        for k, v in root['children'].items():
+            fixup_tree_item(v, label, value, state)
+
+
+    output = {'/': tree['/']}
+
+    del tree['/']
+
+    for k, v in tree.items():
+        fixup_tree_item(output['/'], k, v, state)
+
+    return output
+
+
+def resolve_refs(tree):
+    def build_path_traverse(root, ref):
+        if not root or not 'label' in root:
+            return False
+
+        if ref == root['label']:
+            return True
+
+        for k, v in root['children'].items():
+            path = build_path_traverse(v, ref)
+            if path:
+                if path == True:
+                    return [k]
+
+                return [k, path]
+
+
+    def flatten(lst):
+        for item in lst:
+            if isinstance(item, str):
+                yield item
+            else:
+                yield from flatten(item)
+
+
+    def build_path(root, ref):
+        built = build_path_traverse(root, ref)
+        if not built:
+            return '&%s' % ref
+
+        return '/' + '/'.join(flatten(built))
+
+
+    def resolve_refs_items(root, branch):
+        if not branch:
+            return
+
+        for k, v in branch.get('props', {}).items():
+            if isinstance(v, dict) and 'ref' in v:
+                branch['props'][k] = build_path(root, v['ref'])
+        for k, v in branch.get('children', {}).items():
+            if isinstance(v, dict):
+                resolve_refs_items(root, v)
+
+    resolve_refs_items(tree['/'], tree['/'])
+
+    return tree
+
+
+def flatten(input, output, boot_cpu, include_path):
+    output.write('/dts-v1/;\n')
+
+    tree = devicetree.parse_file(input, include_path=include_path)
+
+    tree = fixup_tree(tree)
+
+    tree = resolve_refs(tree)
+
+    dump_tree(tree['/'], tree, output)
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        formatter_class=argparse.RawDescriptionHelpFormatter)
+
+    parser.add_argument('-b', '--boot-cpu',
+                        required=True, help="Boot CPU number")
+    parser.add_argument('-I', '--include-path',
+                        required=True, action='append',
+                        help="Add directory to include path")
+    parser.add_argument('-i', '--input-dts',
+                        required=True, help="Input DTS path")
+    parser.add_argument('-o', '--output-dts',
+                        required=True, help="Output DTS path")
+
+    args = parser.parse_args()
+
+    with open(args.input_dts, 'r') as input_file:
+        with open(args.output_dts, 'w') as output_file:
+            return flatten(input_file, output_file,
+                           args.boot_cpu, args.include_path)
+
+if __name__ == '__main__':
+    sys.exit(main())
```

Comment:
I haven't seen sys exit used here before.

I see that flatten doesn't return anything, so sys exist isn't doing anything is it?

Reply:
It returns `None`, which ends up being 0.  I use this idiom as it's often easier to just return from main() (like any C program) rather than call sys.exit() in main() directly.  Not a big deal in this case, though.